master/internal/rm/kubernetesrm/pods.go

```diff
@@ -952,35 +978,46 @@ func (p *pods) receiveKillPod(ctx *actor.Context, msg KillTaskPod) {
 		return
 	}
 
-	ctx.Tell(ref, msg)
+	ref.KillTaskPod()
 }
 
-func (p *pods) cleanUpPodHandler(ctx *actor.Context, podHandler *actor.Ref) error {
+func (p *pods) cleanUpPodHandler(podHandler *pod) error {
+	podHandler.finalize()
+
 	podInfo, ok := p.podHandlerToMetadata[podHandler]
 	if !ok {
-		return errors.Errorf("unknown pod handler being deleted %s", podHandler.Address())
+		return errors.Errorf("unknown pod handler being deleted %s", podHandler.podName)
 	}
 
-	name := fmt.Sprintf("%s-priorityclass", podInfo.containerID)
-	_, exists := p.clientSet.SchedulingV1().PriorityClasses().Get(
-		context.TODO(), name, metaV1.GetOptions{})
-	if exists == nil {
-		err := p.clientSet.SchedulingV1().PriorityClasses().Delete(
-			context.TODO(), name, metaV1.DeleteOptions{})
-		if err != nil {
-			ctx.Log().Warnf("Deletion of PriorityClass %s failed.", name)
-		}
-	}
-
-	ctx.Log().WithField("pod", podInfo.podName).WithField(
-		"handler", podHandler.Address()).Infof("de-registering pod handler")
+	p.syslog.WithField("pod", podInfo.podName).WithField(
+		"handler", podHandler.podName).Infof("de-registering pod handler")
 	delete(p.podNameToPodHandler, podInfo.podName)
 	delete(p.podNameToResourcePool, podInfo.podName)
 	delete(p.podNameToContainerID, podInfo.podName)
 	delete(p.containerIDToPodName, podInfo.containerID)
 	delete(p.containerIDToSchedulingState, podInfo.containerID)
 	delete(p.podHandlerToMetadata, podHandler)
 
+	// launch this work async, since we hold the lock and it does API calls.
+	p.wg.Go(func(ctx context.Context) {
+		name := fmt.Sprintf("%s-priorityclass", podInfo.containerID)
+		_, exists := p.clientSet.
+			SchedulingV1().
+			PriorityClasses().
+			Get(ctx, name, metaV1.GetOptions{})
+		if exists != nil {
+			return
+		}
```

Comment:
It'd be nice if we could skip this Get operation and do a Delete. Downside of that, here, is that we log a failed delete which *isn't* a "not found" error.

Digging through the k8s code (I really wish it was better documented), I came across [`errors.IsNotFound`](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsNotFound); if that can successfully detect the case that the PriorityClass doesn't exist when we `Delete()`, we could remove an API call.

Kind of a nit, though, since this works.

Reply:
agree, i noticed this too. just didn't take the time to dig in mid-refactor. thanks.